File: Algorithms/0677_Map_Sum_Pairs/Java/MapSumPairsSolution1.java

// Space: O(n)
// Time: O(n)

class MapSum {

    private Map<String, Integer> numberMap;
    private Map<String, List<String>> prefixMap;

    public MapSum() {
        numberMap = new HashMap<>();
        prefixMap = new HashMap<>();
    }

    public void insert(String key, int val) {
        if (!numberMap.containsKey(key)) {
            for (int i = 1; i < key.length() + 1; i++) {
                prefixMap.computeIfAbsent(key.substring(0, i), (k) -> new ArrayList<>()).add(key);
            }
        }
        numberMap.put(key, val);
    }

    public int sum(String prefix) {
        int res = 0;
        if (prefixMap.containsKey(prefix)) {
            for (String s : prefixMap.get(prefix)) {
                res += numberMap.get(s);
            }
        }
        return res;
    }
}

File: Algorithms/0677_Map_Sum_Pairs/Python/Map_Sum_Pairs_Solution_1.py

# Space: O(n)
# Time: O(n)


import collections


class MapSum:

    def __init__(self):
        self.number_dict = {}
        self.prefix_dict = collections.defaultdict(list)

    def insert(self, key: str, val: int) -> None:
        if key not in self.number_dict:
            for i in range(1, len(key) + 1):
                self.prefix_dict[key[:i]].append(key)
        self.number_dict[key] = val

    def sum(self, prefix: str) -> int:
        return sum(self.number_dict[key] for key in self.prefix_dict[prefix])





